PyInstaller/loader/pyimod03_importers.py

```diff
@@ -491,6 +490,123 @@ def get_filename(self, fullname):
                 fullname.replace('.', pyi_os_path.os_sep) + '.pyc')
         return filename
 
+    def find_spec(self, fullname, path=None, target=None):
+        """
+        PEP-451 finder.find_spec() method for the ``sys.meta_path`` hook.
+
+        fullname     fully qualified name of the module
+        path         None for a top-level module, or package.__path__ for submodules or subpackages.
+
+        Finders must return ModuleSpec objects when find_spec() is called. This new method replaces find_module()
+        and find_loader() (in the PathEntryFinder case). If a loader does not have find_spec(), find_module() and
+        find_loader() are used instead, for backward-compatibility.
+        """
+        # Acquire the interpreter's import lock for the current thread. This
+        # lock should be used by import hooks to ensure thread-safety when
+        # importing modules.
+
+        filename = None  # None means - no module found in this importer.
+
+        if fullname in self.toc:
+            # Tell the import machinery to use self.load_module() to load the module.
+            filename = fullname
+            trace("import %s # PyInstaller PYZ", fullname)
+        elif path is not None:
+            # Try to handle module.__path__ modifications by the modules themselves
+            # Reverse the fake __path__ we added to the package module to a
+            # dotted module name and add the tail module from fullname onto that
+            # to synthesize a new fullname
+            modname = fullname.split('.')[-1]
+
+            for p in path:
+                p = p.replace(SYS_PREFIX, "")
+                parts = p.split(pyi_os_path.os_sep)
+                if not len(parts): continue
+                if not parts[0]:
+                    parts = parts[1:]
+                parts.append(modname)
+                path = ".".join(parts)
+                if path in self.toc:
+                    filename = path
+                    trace("import %s as %s # PyInstaller PYZ (__path__ override: %s)",
+                          path, fullname, p)
+                    break
+
+        if filename is None:
+            trace("# %s not found in PYZ", fullname)
+            return None
+
+        is_pkg, bytecode = self._pyz_archive.extract(filename)
+        if is_pkg:
+            origin = pyi_os_path.os_path_join(pyi_os_path.os_path_join(SYS_PREFIX, filename.replace('.',
+                                                                                                    pyi_os_path.os_sep)),
+                                              '__init__.pyc')
+        else:
+            origin = pyi_os_path.os_path_join(SYS_PREFIX, filename.replace('.', pyi_os_path.os_sep) + '.pyc')
+
+        if not hasattr(self, 'module_code'):
+            self.module_code = {}
+
+        self.module_code[fullname] = bytecode
```

Comment:
> ROFL. You would not even be able to implement PEP 302 protocol, since it already is. SCNR.

Yes and no. The the current PEP 302 methods currently call `extract`, which as I said may be costly. When I said "implement," I _really_ meant re-implement.

> Anyway, I 'm in strong favour of removing this. find_spec is a finder method, not a loader one. So this should be about finding and the loading will be done later (maybe in exec_module).

Yes, and in fact that is the way it is supposed to work but as I said, I avoided it because of `extract`. I feel like I'm repeating myself now.